R/bioc-standalone.R

```diff
@@ -0,0 +1,247 @@
+#' Tools for Bioconductor versions and repositories
+#'
+#' \section{API:}
+#'
+#' ```
+#' get_yaml_config(forget = FALSE)
+#' set_yaml_config(text)
+#'
+#' get_release_version(forget = FALSE)
+#' get_devel_version(forget = FALSE)
+#'
+#' get_version_map(forget = FALSE)
+#' get_matching_bioc_version(r_version = getRversion(), forget = FALSE)
+#' get_bioc_version(r_version = getRversion(), forget = FALSE)
+#'
+#' get_repos(bioc_version = "auto", forget = FALSE)
+#' ```
+#'
+#' * `forget`: Whether to forget the cached version of the Bioconductor
+#'   config YAML file and download it again.
+#' * 
+#' 
+#' 
+#' \section{NEWS:}
+#' * 2019-05-30 First version in remotes.
+#'
+#' 
+#' @name bioconductor
+#' @keywords internal
+#' @noRd
+NULL
+
+
+bioconductor <- local({
+
+  # -------------------------------------------------------------------
+  # Configuration that does not change often
+
+  config_url <- "https://bioconductor.org/config.yaml"
+
+  builtin_map <- list(
+    "2.1"  = package_version("1.6"),
+    "2.2"  = package_version("1.7"),
+    "2.3"  = package_version("1.8"),
+    "2.4"  = package_version("1.9"),
+    "2.5"  = package_version("2.0"),
+    "2.6"  = package_version("2.1"),
+    "2.7"  = package_version("2.2"),
+    "2.8"  = package_version("2.3"),
+    "2.9"  = package_version("2.4"),
+    "2.10" = package_version("2.5"),
+    "2.11" = package_version("2.6"),
+    "2.12" = package_version("2.7"),
+    "2.13" = package_version("2.8"),
+    "2.14" = package_version("2.9"),
+    "2.15" = package_version("2.11"),
+    "3.0"  = package_version("2.13"),
+    "3.1"  = package_version("3.0"),
+    "3.2"  = package_version("3.2"),
+    "3.3"  = package_version("3.4"),
+    "3.4"  = package_version("3.6"),
+    "3.5"  = package_version("3.8")
+  )
+
+  # -------------------------------------------------------------------
+  # Cache
+
+  devel_version <- NULL
+  release_version <- NULL
+  version_map <- NULL
+  yaml_config <- NULL
+
+  # -------------------------------------------------------------------
+  # API
+
+  get_yaml_config <- function(forget = FALSE) {
+    if (forget || is.null(yaml_config)) {
+      new <- tryCatch(readLines(config_url), error = function(x) x)
```

Comment:
How does `readLines()` handle URLs -- does it just invoke the moral equivalent of `utils::download.file()`? (just in case a user relies on e.g. `curl` + having things like `http_proxy` set)

Reply:
Good point. It calls `file()` internally, which uses the `url.default` option. We'll have a "downloader" standalone file soon, and then this bioc file will use that. Would you prefer to use `download.file()` directly until then, instead of `file()`? I'll start writing the downloader code soon, buy it won't make it in the next remotes release.

Reply:
I think `download.file()` would be a better choice for now.

Reply:
OK, I'll change it to that.